### PR TITLE
sync labels

### PR DESCRIPTION
**Pull request by chriku**

### CODE REVIEW

File: github/src/main/kotlin/gropius/sync/github/NodeSourcerer.kt [hunk 1]
```diff
@@ -81,20 +80,19 @@ class NodeSourcerer(
     }
 
     /**
-     * Create an issuebody for the given issue is in the database
-     * @param info The issue to created this body for
+     * Preapre an IssueBody for the given issue is in the database
```
Comment (nk-coding) [inline github/src/main/kotlin/gropius/sync/github/NodeSourcerer.kt:83]:
```suggestion
     * Prepare an IssueBody for the given issue is in the database
```

File: github/src/main/kotlin/gropius/sync/github/Outgoing.kt [hunk 2]
```diff
@@ -184,6 +186,82 @@ class Outgoing(
...
     }
 
+    /**
+     * Mutate an AddedLabelEvent upto GitHub
+     * @param imsProjectConfig active config
+     * @param issueInfo info of the issue containing the timeline item
+     * @param userList users that have contributed to the event
+     * @return List of functions that contain the actual mutation executors
+     */
+    private suspend fun githubAddLabel(
```
Comment (nk-coding) [inline github/src/main/kotlin/gropius/sync/github/Outgoing.kt:196]: too long, maybe extract both if cases into its own methods?

File: github/src/main/kotlin/gropius/sync/github/Outgoing.kt [hunk 3]
```diff
@@ -205,21 +283,108 @@ class Outgoing(
             }
         }
         val collectedMutations = mutableListOf<suspend () -> Unit>()
-        if (relevantTimeline.last() is ReopenedEvent) {
-            collectedMutations += githubReopenIssue(
-                imsProjectConfig,
+        if (shouldSyncType<ReopenedEvent, ClosedEvent>(
```
Comment (nk-coding) [inline github/src/main/kotlin/gropius/sync/github/Outgoing.kt:286]: too long with these changes

File: github/src/main/kotlin/gropius/sync/github/Outgoing.kt [hunk 4]
```diff
@@ -205,21 +283,108 @@ class Outgoing(
...
+    }
+
+    /**
+     * Mutate the labels of an issue upto GitHub
+     * @param imsProjectConfig active config
+     * @param issueInfo info of the issue containing the timeline item
+     * @param timeline TimelineItems for this issue
+     * @return List of functions that contain the actual mutation executors
+     */
+    private suspend fun pushLabels(
```
Comment (nk-coding) [inline github/src/main/kotlin/gropius/sync/github/Outgoing.kt:341]: too long

File: github/src/main/kotlin/gropius/sync/github/Outgoing.kt [hunk 5]
```diff
@@ -184,6 +186,82 @@ class Outgoing(
         }
     }
 
+    /**
+     * Mutate an AddedLabelEvent upto GitHub
+     * @param imsProjectConfig active config
+     * @param issueInfo info of the issue containing the timeline item
+     * @param userList users that have contributed to the event
```
Comment (nk-coding) [inline github/src/main/kotlin/gropius/sync/github/Outgoing.kt:193]: missing doc for label

File: github/src/main/kotlin/gropius/sync/github/Outgoing.kt [hunk 6]
```diff
@@ -184,6 +186,82 @@ class Outgoing(
...
+                    val response =
+                        client.mutation(MutateAddLabelMutation(issueInfo.githubId, labelId.githubId)).execute()
+                    val item =
+                        response.data?.addLabelsToLabelable?.labelable?.asIssue()?.timelineItems?.nodes?.lastOrNull()
+                    if (item != null) {
+                        incoming.handleTimelineEvent(imsProjectConfig, issueInfo, item)
+                    }
+                }
+            } else {
+                listOf();
```
Comment (nk-coding) [inline github/src/main/kotlin/gropius/sync/github/Outgoing.kt:214]: maybe prefer emptyList

File: github/src/main/kotlin/gropius/sync/github/Outgoing.kt [hunk 7]
```diff
@@ -184,6 +186,82 @@ class Outgoing(
...
+                }
+            }
+        }
+    }
+
+    /**
+     * Mutate an RemovedLabelEvent upto GitHub
+     * @param imsProjectConfig active config
+     * @param issueInfo info of the issue containing the timeline item
+     * @param userList users that have contributed to the event
```
Comment (nk-coding) [inline github/src/main/kotlin/gropius/sync/github/Outgoing.kt:243]: missing doc for label

File: github/src/main/kotlin/gropius/sync/github/Outgoing.kt [hunk 8]
```diff
@@ -205,21 +283,108 @@ class Outgoing(
...
                 finalBlock.map { it.lastModifiedBy().value })
         }
-        if (relevantTimeline.last() is ClosedEvent) {
-            collectedMutations += githubCloseIssue(
-                imsProjectConfig,
+        if (shouldSyncType<ClosedEvent, ReopenedEvent>(
+                finalBlock, relevantTimeline, false
+            )
+        ) {
+            collectedMutations += githubCloseIssue(imsProjectConfig,
```
Comment (nk-coding) [inline github/src/main/kotlin/gropius/sync/github/Outgoing.kt:298]: you sure the autoformatter did this?

File: github/src/main/kotlin/gropius/sync/github/Outgoing.kt [hunk 9]
```diff
@@ -205,21 +283,108 @@ class Outgoing(
...
+     */
+    private suspend fun pushLabels(
+        imsProjectConfig: IMSProjectConfig, issueInfo: IssueInfo, timeline: List<TimelineItem>
+    ): List<suspend () -> Unit> {
+        val groups = timeline.filter { (it is AddedLabelEvent) || (it is RemovedLabelEvent) }.groupBy {
+            when (it) {
+                is AddedLabelEvent -> {
+                    it.addedLabel().value
+                }
+
```
Comment (nk-coding) [inline github/src/main/kotlin/gropius/sync/github/Outgoing.kt:349]: maybe remove these empty lines

**Review by nk-coding — CHANGES_REQUESTED** [review]

File: github/src/main/kotlin/gropius/sync/github/Outgoing.kt (hunk 9 above)
Comment (chriku) [reply github/src/main/kotlin/gropius/sync/github/Outgoing.kt:349]: I CAN'T! FORMATTER REVERTS THIS

File: github/src/main/kotlin/gropius/sync/github/Outgoing.kt (hunk 8 above)
Comment (chriku) [reply github/src/main/kotlin/gropius/sync/github/Outgoing.kt:298]: oscillates between this and 
```kt
            collectedMutations += githubCloseIssue(
                imsProjectConfig,
                issueInfo,
                finalBlock.map { it.lastModifiedBy().value })
```
every few formats

File: github/src/main/kotlin/gropius/sync/github/Outgoing.kt (hunk 5 above)
Comment (chriku) [reply github/src/main/kotlin/gropius/sync/github/Outgoing.kt:193]: f4cea4b1d0191373ae90c82a78fcdbe78e5cf47d

File: github/src/main/kotlin/gropius/sync/github/Outgoing.kt (hunk 7 above)
Comment (chriku) [reply github/src/main/kotlin/gropius/sync/github/Outgoing.kt:243]: f4cea4b1d0191373ae90c82a78fcdbe78e5cf47d

File: github/src/main/kotlin/gropius/sync/github/Outgoing.kt (hunk 6 above)
Comment (chriku) [reply github/src/main/kotlin/gropius/sync/github/Outgoing.kt]: 95e766e888b622bce1397c183141025ead868dcb

File: github/src/main/kotlin/gropius/sync/github/Outgoing.kt (hunk 2 above)
Comment (chriku) [reply github/src/main/kotlin/gropius/sync/github/Outgoing.kt:198]: 90cc10cffa55b80cdf692b34b00baa1c425b3577

File: github/src/main/kotlin/gropius/sync/github/Outgoing.kt (hunk 3 above)
Comment (chriku) [reply github/src/main/kotlin/gropius/sync/github/Outgoing.kt:330]: a73e319646f896cd415f11db6e96fc5bd9b29607

File: github/src/main/kotlin/gropius/sync/github/Outgoing.kt (hunk 4 above)
Comment (chriku) [reply github/src/main/kotlin/gropius/sync/github/Outgoing.kt:385]: b6657f04f5a2753892bf920dc80b7bd4b08681f9

File: github/src/main/kotlin/gropius/sync/github/Outgoing.kt (hunk 9 above)
Comment (chriku) [reply github/src/main/kotlin/gropius/sync/github/Outgoing.kt]: b6657f04f5a2753892bf920dc80b7bd4b08681f9

File: github/src/main/kotlin/gropius/sync/github/Outgoing.kt (hunk 8 above)
Comment (nk-coding) [reply github/src/main/kotlin/gropius/sync/github/Outgoing.kt:342]: strange

**Review by nk-coding — APPROVED** [review]

File: github/src/main/kotlin/gropius/Application.kt [hunk 10]
```diff
@@ -55,8 +55,12 @@ class Application : CommandLineRunner {
...
 
     override fun run(vararg args: String?) {
-        runBlocking {
-            syncSelector.sync()
+        try {
+            runBlocking {
+                syncSelector.sync()
+            }
+        } finally {
+            exitProcess(0)
```
Comment (ModellbahnFreak) [inline github/src/main/kotlin/gropius/Application.kt:63]: Maybe an exit code with semantic meaning (or at least non 0 when error?)

**Review by ModellbahnFreak — CHANGES_REQUESTED** [review]

File: github/src/main/kotlin/gropius/Application.kt (hunk 10 above)
Comment (chriku) [reply github/src/main/kotlin/gropius/Application.kt]: 21296399359923d19257bbc271b41f71077fc218

Comment (ModellbahnFreak) [reply github/src/main/kotlin/gropius/Application.kt]: Closing and appriving asap 😉

**Review by ModellbahnFreak — APPROVED** [review]

**Pull request merged**